nixos/modules/services/logging/journaldriver.nix

```diff
@@ -0,0 +1,117 @@
+# Logging.
+#
+# It can be enabled without extra configuration when running on GCP.
+# On machines hosted elsewhere, the other configuration options need
+# to be set.
+#
+# For further information please consult the documentation in the
+# upstream repository at: https://github.com/aprilabank/journaldriver/
+
+{ config, lib, pkgs, ...}:
+
+with lib; let cfg = config.services.journaldriver;
+envVars = [
+  { name = "RUST_LOG"; value = cfg.logLevel; }
+  { name = "LOG_NAME"; value = cfg.logName; }
+  { name = "LOG_STREAM"; value = cfg.logStream; }
+  { name = "GOOGLE_CLOUD_PROJECT"; value = cfg.googleCloudProject; }
+  { name = "GOOGLE_APPLICATION_CREDENTIALS"; value = cfg.applicationCredentials; }
+];
+
+# Assemble an environment in which all values that have not been
+# specified explicitly are left unset:
+environment = with builtins; listToAttrs (filter (v: v.value != null) envVars);
+in {
+  options.services.journaldriver = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable journaldriver to forward journald logs to
+        Stackdriver Logging.
+      '';
+    };
+
+    logLevel = mkOption {
+      type = types.string;
+      default = "info";
+      description = ''
+        Log level at which journaldriver logs its own output.
+      '';
+    };
+
+    logName = mkOption {
+      type        = with types; nullOr string;
+      default     = null;
+      description = ''
+        Configures the name of the target log in Stackdriver Logging.
+        This option can be set to, for example, the hostname of a
+        machine to improve the user experience in the logging
+        overview.
+      '';
+    };
+
+    googleCloudProject = mkOption {
+      type        = with types; nullOr string;
+      default     = null;
+      description = ''
+        Configures the name of the Google Cloud project to which to
+        forward journald logs.
+
+        This option is required on non-GCP machines, but should not be
+        set on GCP instances.
+      '';
+    };
+
+    logStream = mkOption {
+      type        = with types; nullOr string;
+      default     = null;
+      description = ''
+        Configures the name of the Stackdriver Logging log stream into
+        which to write journald entries.
+
+        This option is required on non-GCP machines, but should not be
+        set on GCP instances.
+      '';
+    };
+
+    applicationCredentials = mkOption {
+      type = with types; nullOr string;
+      default = null;
+      description = ''
+        Path to the service account private key (in JSON-format) used
+        to forward log entries to Stackdriver Logging on non-GCP
+        instances.
+
+        This option is required on non-GCP machines, but should not be
+        set on GCP instances.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.extraUsers.journaldriver = {
+      isSystemUser = true;
```

Comment:
Why is this set to `true`?

Reply:
As far as I understand the differentiation this is only to indicate that the user account is not used by a human, which results in the UID of the account being allocated in a different range.

Reply:
Hmm, I'm not entirely sure, there are a couple services that use this, but others set a fixed id in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/ids.nix#L36. I hope somebody else can comment on this

Reply:
The comment at the top of `ids.nix` sounds to me like the preferred way is to not hardcode an ID unless necessary.

It may be possible to use the `DynamicUser=` option in systemd. Journaldriver needs to have a persistent file to store its cursor position in, this should be possible to achieve with `StateDirectory=` in systemd - it's not obvious how to grant the dynamic user membership in the journal access group though.

Reply:
Update:

This configuration seems to work without specifying a user at all:

```
      serviceConfig = {
        Restart        = "always";
        DynamicUser    = true;
        StateDirectory = "journaldriver";

        # This group is required for accessing journald.
        SupplementaryGroups = "systemd-journal";
      };
```